### DISCUSSION TITLE
CTKEntry how to submit without button

### DISCUSSION THREAD

**Discussion by pinkurauchin** [TomSchimansky/CustomTkinter#928] [Q&A] [answered]
Hi!
First of all, thank you for your hard work, this is such an amazing way to have a good-look application :)

<img width="157" alt="image" src="https://user-images.githubusercontent.com/93226265/209162817-373e33af-ee89-48d7-990f-18b7c31dceb9.png">

The question is,
I have 2 Entries, and i want to execute a def after finish to typing, for example when change the focus.
The most important thing, is that i dont want a button to submit the info.
I tried so many ways, and see information of tkinter and ttk, but i cannot do it in any ways

Thanks for help :)

**Comment by Akascape** [comment, marked as answer]
@pinkurauchin You have to use `tkinter variables` with `trace_add` method.
Here is an example of multiplying two numbers which gives the result by just entering the numbers.

![Screenshot](https://user-images.githubusercontent.com/89206401/209208959-c4af5191-f84b-4909-92fc-de08f1e495ad.jpg)
## CODE
```python
import tkinter
import customtkinter

customtkinter.set_appearance_mode("System") 
customtkinter.set_default_color_theme("blue")

HEIGHT = 200
WIDTH = 300
                
app = customtkinter.CTk()
app.title("Dynamic Entry")
app.geometry((f"{WIDTH}x{HEIGHT}"))

label_1 = customtkinter.CTkLabel(master=app,text="Multiply 2 numbers")
label_1.pack(pady=10, padx=10)

def calculate(var, index, mode):
    if entry_1.get()!="" and entry_2.get()!="":
        label_2.configure(text="Total: "+str(int(entry_1.get())*int(entry_2.get())))
    else:
        label_2.configure(text="Total: --")
        
def only_numbers(char):
    #Validate for numbers only
    if ((char.isdigit()) or (char=="")):
        return True
    else:
        return False
    
validation = app.register(only_numbers)

# Make two Interget variables
var1=tkinter.StringVar() 
var2=tkinter.StringVar()
var1.set("0")
var2.set("0")

entry_1 = customtkinter.CTkEntry(master=app, textvariable=var1, validate='key', validatecommand=(validation, '%P'))
entry_1.pack(pady=10, padx=10)

entry_2 = customtkinter.CTkEntry(master=app, textvariable=var2, validate='key', validatecommand=(validation, '%P'))
entry_2.pack(pady=10, padx=10)

label_2 = customtkinter.CTkLabel(master=app,text="Total: 0")
label_2.pack(pady=10, padx=10)

# Trace the variables
var1.trace_add('write', calculate)
var2.trace_add('write', calculate)

app.mainloop()
```

**Reply by pinkurauchin** [reply]
Hi! Thanks for your fast response and help :)
The problem is that i dont want to execute the calculate after write a number,
if i write in the entry "23", it executes after 2, and after 3, and i just want to do it, after finish typing all the numbers, for example when it changes the focus to other widget :)

thank you!

**Reply by Akascape** [reply]
@pinkurauchin It's easy, you can add conditions like this:
```python
def calculate(var, index, mode):
    if not len(entry_1.get())>=2: # proceed only if entry is more than 2
        return
   ...
```

Or you can also use a bind method, it will only display the result when it changes the focus to other widgets:

```python
import tkinter
import customtkinter

app = customtkinter.CTk()
app.title("Dynamic Entry")
app.geometry("300x200")

label_1 = customtkinter.CTkLabel(master=app,text="Multiply 2 numbers")
label_1.pack(pady=10, padx=10)

def calculate():
    if entry_1.get()!="" and entry_2.get()!="":
        label_2.configure(text="Total: "+str(int(entry_1.get())*int(entry_2.get())))
    else:
        label_2.configure(text="Total: --")
        
def only_numbers(char):
    #Validate for numbers only
    if ((char.isdigit()) or (char=="")):
        return True
    else:
        return False
    
validation = app.register(only_numbers)

# Make two Interget variables
var1=tkinter.StringVar() 
var2=tkinter.StringVar()
var1.set("0")
var2.set("0")

entry_1 = customtkinter.CTkEntry(master=app, textvariable=var1, validate='key', validatecommand=(validation, '%P'))
entry_1.pack(pady=10, padx=10)

entry_2 = customtkinter.CTkEntry(master=app, textvariable=var2, validate='key', validatecommand=(validation, '%P'))
entry_2.pack(pady=10, padx=10)

label_2 = customtkinter.CTkLabel(master=app,text="Total: 0")
label_2.pack(pady=10, padx=10)

app.bind("<Button-1>", lambda e: calculate()) # call the function only after clicking somewhere

app.mainloop()                  
```

**Reply by pinkurauchin** [reply]
Omg, it works perfectly, ty so much, i just have a bit questions more

How it works bind? It's afert clicking somewhere else, but, you can do it with other widget, and not with the all program? 
Something like:

`entry_2.bind("<Button-1>", lambda e: calculate())`
and other one for the other entry, because now i want 2 methods :)

and the other question is: what is exactly "<Button-1>", or whats does it means, why is there, which other options are...

Thank you so much! :)

**Reply by pinkurauchin** [reply]
To give you more info:
<img width="154" alt="image" src="https://user-images.githubusercontent.com/93226265/209558411-71786951-d6bf-4cb2-8f58-02c6bdb63dc1.png">


This is the menu that i'm doing, the problem is that always execute the command when i press in the background of the app, its a frame.

So, i just want to execute the method when i change the data in the boxes, but i need 1 method for eachone.

Right now, i have just one method which executes when i press anywhere, and it changes the data everytime, i just want it, when really really change it, for example when i change the 45 to 89.
And obviously, i need a method for eachone box, so, 2 methods :)
Maybe this is not the best way, but they want me to do in this way hahahah
Ty!

**Reply by pinkurauchin** [reply]
The code:

```
def changedata():
 ...

minutes = customtkinter.CTkEntry(master=mod1_frame, textvariable=minutes_var, validate='key', placeholder_text=minutes_var.get(),
                                         validatecommand=(validation, '%P'))
minutes.pack(pady=10, padx=10)

seconds = customtkinter.CTkEntry(master=mod1_frame, textvariable=seconds_var, validate='key', placeholder_text=seconds_var.get(),
                                         validatecommand=(validation, '%P'))
seconds.pack(pady=10, padx=10)

...

mod1_frame.bind("<Button-1>", lambda e: changedata())
```

**Comment by Akascape** [comment]
@pinkurauchin 
>How it works bind?

Bind is a method to link any key or event with a command. Here `<button-1>` is the left mouse click. You can find many other bind events [here](https://stackoverflow.com/a/32289245/16813096).

>  It's afert clicking somewhere else, but, you can do it with other widget, and not with the all program?
> Something like:
> `entry_2.bind("<Button-1>", lambda e: calculate())`

Yes, you can do that. In this case the calculate function will only be called when you click the entry box.
> and the other question is: what is exactly ""

`""` means that the string is empty. We can put conditions (if-else) that if the string is empty then don't proceed the command.

> This is the menu that i'm doing, the problem is that always execute the command when i press in the background of the app, its a frame.

There is another improved method you can use in which the command will only proceed if you have changed the values in box:

```python
old_var1 = "0"
old_var2 = "0"

def calculate():
    global old_var1, old_var2
    if old_var1==var1.get() and old_var2==var2.get():
        return # if old values are not changes then return

    if entry_1.get()!="" and entry_2.get()!="":
        label_2.configure(text="Total: "+str(int(entry_1.get())*int(entry_2.get())))
    else:
        label_2.configure(text="Total: --")
        
    # update the old values
    old_var1 = var1.get()
    old_var2 = var2.get()
```

**Reply by pinkurauchin** [reply]
OMG, finally i have it all fixed, thank you so much for your help :)
